src/query/storage/m3/storage.go

```diff
@@ -329,58 +329,103 @@ func (s *m3storage) CompleteTags(
 	default:
 	}
 
-	// TODO: instead of aggregating locally, have the DB aggregate it before
-	// sending results back.
 	fetchQuery := &storage.FetchQuery{
 		TagMatchers: query.TagMatchers,
-		// NB: complete tags matches every tag from the start of time until now
-		Start: time.Time{},
-		End:   time.Now(),
 	}
 
-	results, cleanup, err := s.SearchCompressed(ctx, fetchQuery, options)
-	defer cleanup()
+	m3query, err := storage.FetchQueryToM3Query(fetchQuery, s.conversionCache)
 	if err != nil {
 		return nil, err
 	}
 
-	accumulatedTags := storage.NewCompleteTagsResultBuilder(query.CompleteNameOnly)
-	// only filter if there are tags to filter on.
-	filtering := len(query.FilterNameTags) > 0
-	for _, elem := range results {
-		it := elem.Iter
-		tags := make([]storage.CompletedTag, 0, it.Len())
-		for i := 0; it.Next(); i++ {
-			tag := it.Current()
-			name := tag.Name.Bytes()
-			if filtering {
-				found := false
-				for _, filterName := range query.FilterNameTags {
-					if bytes.Equal(filterName, name) {
-						found = true
-						break
-					}
+	aggOpts := storage.FetchOptionsToAggregateOptions(options, query)
+
+	var (
+		namespaces      = s.clusters.ClusterNamespaces()
+		accumulatedTags = storage.NewCompleteTagsResultBuilder(query.CompleteNameOnly)
+		multiErr        syncMultiErrs
+		wg              sync.WaitGroup
+	)
+
+	if len(namespaces) == 0 {
+		return nil, errNoNamespacesConfigured
+	}
+
+	// TODO: (arnikola): implement a better lifecycle for incoming aggregated tag
```

Comment:
do you want to do this in this PR? 

super nit: `TODO(arnikola): ` instead of `TODO: (arnikola)`

Reply:
I think there's some low hanging fruit gains for query in general, but probably best not to add them here and risk weird pooling issues for now